File: ts/src/mcp/runtime-session-tools.ts
import { z } from "zod";

import { RuntimeSessionEventStore } from "../session/runtime-events.js";
import { runtimeSessionIdForRun } from "../session/runtime-session-ids.js";
import {
  readRuntimeSessionById,
  readRuntimeSessionByRunId,
  readRuntimeSessionSummaries,
  type RuntimeSessionReadStore,
} from "../session/runtime-session-read-model.js";
import { buildRuntimeSessionTimeline } from "../session/runtime-session-timeline.js";

interface JsonToolResponse {
  content: Array<{
    type: "text";
    text: string;
  }>;
}

type McpToolRegistrar = {
  tool: (...args: any[]) => unknown;
};

type RuntimeSessionClosableReadStore = RuntimeSessionReadStore & {
  close?: () => void;
};

interface RuntimeSessionToolInternals {
  createEventStore(dbPath: string): RuntimeSessionClosableReadStore;
}

const defaultInternals: RuntimeSessionToolInternals = {
  createEventStore: (dbPath) => new RuntimeSessionEventStore(dbPath),
};

const ListRuntimeSessionsArgsSchema = z.object({
  limit: z.number().int().default(50).describe("Max runtime sessions to return"),
});
type ListRuntimeSessionsArgs = z.infer<typeof ListRuntimeSessionsArgsSchema>;

const GetRuntimeSessionArgsSchema = z.object({
  sessionId: z.string().optional().describe("Runtime session ID"),
  runId: z.string().optional().describe("Run ID for the run-scoped runtime session"),
});
type GetRuntimeSessionArgs = z.infer<typeof GetRuntimeSessionArgsSchema>;

export function buildRuntimeSessionIdentifierRequiredPayload(): { error: string } {
  return { error: "get_runtime_session requires sessionId or runId" };
}

export function buildRuntimeSessionIdentifierConflictPayload(): { error: string } {
  return { error: "get_runtime_session accepts only one of sessionId or runId" };
}

export function buildRuntimeSessionNotFoundPayload(sessionId: string): {
  error: string;
  session_id: string;
} {
  return { error: "Runtime session not found", session_id: sessionId };
}

export function registerRuntimeSessionTools(
  server: McpToolRegistrar,
  opts: {
    dbPath?: string;
    store?: RuntimeSessionReadStore;
    internals?: Partial<RuntimeSessionToolInternals>;
  },
): void {
  const internals: RuntimeSessionToolInternals = {
    ...defaultInternals,
    ...opts.internals,
  };

  server.tool(
    "list_runtime_sessions",
    "List recent runtime-session event logs",
    ListRuntimeSessionsArgsSchema.shape,
    async (args: ListRuntimeSessionsArgs) =>
      withRuntimeSessionStore(opts, internals, (store) =>
        jsonText(
          {
            sessions: readRuntimeSessionSummaries(store, {
              limit: args.limit ?? 50,
            }),
          },
          2,
        ),
      ),
  );

  server.tool(
    "get_runtime_session",
    "Read a runtime-session event log by session id or run id",
    GetRuntimeSessionArgsSchema.shape,
    async (args: GetRuntimeSessionArgs) =>
      withRuntimeSessionStore(opts, internals, (store) => {
        const sessionId = cleanIdentifier(args.sessionId);
        const runId = cleanIdentifier(args.runId);
        if (!sessionId && !runId) {
          return jsonText(buildRuntimeSessionIdentifierRequiredPayload());
        }
        if (sessionId && runId) {
          return jsonText(buildRuntimeSessionIdentifierConflictPayload());
        }

        const log = sessionId
          ? readRuntimeSessionById(store, sessionId)
          : readRuntimeSessionByRunId(store, runId);
        const resolvedSessionId = sessionId || runtimeSessionIdForRun(runId);
        if (!log) {
          return jsonText(buildRuntimeSessionNotFoundPayload(resolvedSessionId));
        }
        return jsonText(log.toJSON(), 2);
      }),
  );

  server.tool(
    "get_runtime_session_timeline",
    "Read an operator-facing runtime-session timeline by session id or run id",
    GetRuntimeSessionArgsSchema.shape,
    async (args: GetRuntimeSessionArgs) =>
      withRuntimeSessionStore(opts, internals, (store) => {
        const result = loadRuntimeSessionFromArgs(store, args);
        if ("error" in result) {
          return jsonText(result);
        }
        return jsonText(buildRuntimeSessionTimeline(result.log), 2);
      }),
  );
}

type RuntimeSessionLookupResult =
  | { log: NonNullable<ReturnType<typeof readRuntimeSessionById>> }
  | { error: string; session_id?: string };

function loadRuntimeSessionFromArgs(
  store: RuntimeSessionReadStore,
  args: GetRuntimeSessionArgs,
): RuntimeSessionLookupResult {
  const sessionId = cleanIdentifier(args.sessionId);
  const runId = cleanIdentifier(args.runId);
  if (!sessionId && !runId) {
    return buildRuntimeSessionIdentifierRequiredPayload();
  }
  if (sessionId && runId) {
    return buildRuntimeSessionIdentifierConflictPayload();
  }
  const log = sessionId
    ? readRuntimeSessionById(store, sessionId)
    : readRuntimeSessionByRunId(store, runId);
  const resolvedSessionId = sessionId || runtimeSessionIdForRun(runId);
  if (!log) {
    return buildRuntimeSessionNotFoundPayload(resolvedSessionId);
  }
  return { log };
}

function withRuntimeSessionStore(
  opts: {
    dbPath?: string;
    store?: RuntimeSessionReadStore;
  },
  internals: RuntimeSessionToolInternals,
  callback: (store: RuntimeSessionReadStore) => JsonToolResponse,
): JsonToolResponse {
  if (opts.store) {
    return callback(opts.store);
  }
  if (!opts.dbPath) {
    return jsonText({ error: "Runtime session store requires dbPath" });
  }

  const store = internals.createEventStore(opts.dbPath);
  try {
    return callback(store);
  } finally {
    store.close?.();
  }
}

function jsonText(payload: unknown, indent?: number): JsonToolResponse {
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(payload, null, indent),
      },
    ],
  };
}

function cleanIdentifier(value: string | undefined): string {
  return typeof value === "string" ? value.trim() : "";
}

File: autocontext/src/autocontext/mcp/runtime_session_tools.py
from __future__ import annotations

from typing import Any

from autocontext.mcp._base import MtsToolContext
from autocontext.session.runtime_events import RuntimeSessionEventStore
from autocontext.session.runtime_session_ids import runtime_session_id_for_run
from autocontext.session.runtime_session_read_model import (
    read_runtime_session_by_id,
    read_runtime_session_by_run_id,
    read_runtime_session_summaries,
)
from autocontext.session.runtime_session_timeline import (
    read_runtime_session_timeline_by_id,
    read_runtime_session_timeline_by_run_id,
)


def list_runtime_sessions(ctx: MtsToolContext, limit: int = 50) -> dict[str, Any]:
    """List recent runtime-session event logs."""
    store = _event_store(ctx)
    try:
        return {"sessions": read_runtime_session_summaries(store, limit=limit)}
    finally:
        store.close()


def get_runtime_session(
    ctx: MtsToolContext,
    *,
    session_id: str | None = None,
    run_id: str | None = None,
) -> dict[str, Any]:
    """Read a runtime-session event log by session id or run id."""
    resolved = _resolve_runtime_session_identifier(session_id=session_id, run_id=run_id)
    if resolved["error"]:
        return {"error": resolved["error"]}

    store = _event_store(ctx)
    try:
        log = (
            read_runtime_session_by_run_id(store, resolved["run_id"])
            if resolved["run_id"]
            else read_runtime_session_by_id(store, resolved["session_id"])
        )
        if log is None:
            return {"error": "Runtime session not found", "session_id": resolved["session_id"]}
        return log.to_dict()
    finally:
        store.close()


def get_runtime_session_timeline(
    ctx: MtsToolContext,
    *,
    session_id: str | None = None,
    run_id: str | None = None,
) -> dict[str, Any]:
    """Read an operator-facing runtime-session timeline by session id or run id."""
    resolved = _resolve_runtime_session_identifier(session_id=session_id, run_id=run_id)
    if resolved["error"]:
        return {"error": resolved["error"]}

    store = _event_store(ctx)
    try:
        timeline = (
            read_runtime_session_timeline_by_run_id(store, resolved["run_id"])
            if resolved["run_id"]
            else read_runtime_session_timeline_by_id(store, resolved["session_id"])
        )
        if timeline is None:
            return {"error": "Runtime session not found", "session_id": resolved["session_id"]}
        return timeline
    finally:
        store.close()


def _event_store(ctx: MtsToolContext) -> RuntimeSessionEventStore:
    return RuntimeSessionEventStore(ctx.settings.db_path)


def _resolve_runtime_session_identifier(
    *,
    session_id: str | None,
    run_id: str | None,
) -> dict[str, str]:
    clean_session_id = (session_id or "").strip()
    clean_run_id = (run_id or "").strip()
    if bool(clean_session_id) == bool(clean_run_id):
        return {
            "error": "Provide exactly one of session_id or run_id",
            "session_id": "",
            "run_id": "",
        }
    return {
        "error": "",
        "session_id": clean_session_id or runtime_session_id_for_run(clean_run_id),
        "run_id": clean_run_id,
    }
